design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
+
+We may want to introduce a limit to the number of rows that may be returned from these `executeQuery()` methods,
+in order to decrease the likelihood of running out of memory.
+
+### Insert values
+
+A method to insert one row at a time:
+```
+  /**
+   * Insert a single row into the relevant stream/table.
+   *
+   * @param streamName name of stream/table.
+   * @param row the row to insert.
+   * @return a future that completes once the request has been processed.
+   */
+  CompletableFuture<Void> insertInto(String streamName, Map<String, Object> row);
+```
+
+A method to stream inserts (via the `/inserts-stream` endpoint):
+```
+  CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<List<Object>> insertsPublisher);
+}
+```
+where `InsertResponse` is as follows:
+```
+public interface InsertResponse {
+
+  /**
+   * Whether the row was successfully inserted or not.
+   */
+  boolean isSuccessful();
+
+  /**
+   * Unique sequence number for the row in the stream of inserts.
+   */
+  int getSequenceNum();
```

Comment:
So basically the offset? This would not be unique. Do we think we need to expose the partition number, too?

Reply:
This is not the Kafka topic offset. This is the row number in the stream of inserts, meaning the user opens a connection to stream inserts, sends one row, gets an ack with sequence number 1 (I forget if this is 1-indexed or 0-indexed right now, have to check the code), sends another row, gets an ack with sequence number 2, etc.

Reply:
I see. Why do we expose this? Why would the user care?

Reply:
This is so the user can correlate the ack with the insert they sent. E.g. they may want to take some action once they know the insert has been completed ok.

Reply:
What ack? The `InsertResponse` itself tells the user (via `isSuccessful` if the insert was sucessful or not)? I don't see any API where the user would pass in the sequence number of any API with an "ack number" that could be compared to this sequence number? Seems I am missing something?

Reply:
When a user calls `streamInserts(...)`, they'll receive a `Publisher<InsertResponse>` that they can subscribe to in their app code and (asynchronously) take an action whenever a new ack is received. In this way, the user doesn't need to "pass in the sequence number [to] any API" in order to take actions based on acks. That logic would all live in the app code.

Reply:
Acks can come back out of order. so the sequence number is necessary to correlate them with the corresponding inserts.

Reply:
An alternative would be to provide an API, something like the following for streaming inserts:
````
void streamInserts(String streamName, Publisher<Pair<List<Object>, CompletableFuture<Void>>> insertsPublisher);
````
The user would supply the cf for each insert, and would be notified via the cf on completion of the insert. This wouldn't require the user to do any correlation between inserts and acks themselves.

Reply:
> In this way, the user doesn't need to "pass in the sequence number [to] any API" in order to take actions based on acks.

For this case, exposing the sequence number does not seem to be helpful for the user?

> Acks can come back out of order. so the sequence number is necessary to correlate them with the corresponding inserts.

Should we provide a stronger guarantee? Kafka itself also provides strong ordering guarantees. As a user, it seem desirable that the data is not reorder, ie, if I call `streamInsert()` twice (and data goes into the same partition) it should appended in the order of my calls?

I like the idea to just register a callback thought. That is similar to what `KafkaProducer` does.

Reply:
> Should we provide a stronger guarantee? Kafka itself also provides strong ordering guarantees. As a user, it seem desirable that the data is not reorder, ie, if I call streamInsert() twice (and data goes into the same partition)

It's not necessarily going to go in the same partition, it depends on the key. That's why the acks might be in a different order.

Reply:
Sure. On read I don't get any guarantee about the order if the writes went to different partitions. But for this case, the sequence number seems to be even less useful? It's seems to be client side counter? I still don't understand how this is useful? If the `InsertResponse` tells me that the write was successful I am good, otherwise I might retry. Why would I care about this number?

Reply:
The sequence number is necessary because otherwise you don't know what send a particular InsertResponse refers to, as the InsertsResponses can come back in a different order to the sends.

Reply:
Yes, the sequence number of the first insert you send is 0, then next one is 1, the next one 2 etc. That's all you need to correlate. Typically the user would use the sequence_number as the key in a map, so when the response comes back the value for the key can be looked up and some action can be taken.

Reply:
My question was:
```
// fist call
streamInsert(A, B, C) -> 0, 1, 3
// second call
streamInsert(D, E) -> 4, 5 // my original assumption
// or
streamInsert(D, E) -> 0, 1 
```

If it's the former, I am still not sure how user would do the mapping. If it's the later, my original understanding was totally off making the discussion void. However, I would suggest to rename from `sequendNumber` to `index` and update the JavaDocs:
```
/**
  * Index of the the row from the original input `List` provided to `streamInsert()`.
  */
int getInputIndex();
```

"Sequence number" might be rather confusing indicate that it is the position of the record in the stream that the records are appended to. Thoughts?

\cc @vcrfxia

Reply:
It's the latter.
When you call streamInserts you have two streams (reactive streams). 1 - is the stream of inserts going to the server. 2 - is the stream of acks coming back from the server.
Each insert in 1. has a sequence number 0, 1, 2, etc. The sequence number field in the acks refers to the sequence number in 1.
Does that clear things up?

Reply:
Thanks that clears things up. The overloaded terminology of "stream" is really confusing.

Reply:
Reactive streams refers to streams of data sent from publishers to subscribers. gRPC also has streams. HTTP2 has streams too. They're all kind of similar. Then you have streams in ksqlDB/KS. Yep, lots of streams :)